### DISCUSSION TITLE
Custom map routing of `OrchardCore.Users` actions with culture

### DISCUSSION THREAD

**Discussion by DotCat1985** [OrchardCMS/OrchardCore#16477] [Q&A]
I configured my OrchardCore solution for **en** and **it** cultures by using `OrchardCore.Localization` module.
I would like to access to all `OrchardCore.Users` actions only with my custom routes as following:

```
/en/login
/it/login

/en/user-registration
/it/registrazione-utente

...
```
where /en/login will return Login page in English, /it/login will return Login page in Italian, and so on.

Perhaps, how can I map all `OrchardCore.Users` route actions with the example above?

**Comment by Piedone** [comment]
Perhaps you can use `MapAreaControllerRoute()` (see https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0 and example in the OC source) to add a higher prio route in your module/theme's `Startup`?

Also, @hishamco you can help, perhaps?

**Reply by DotCat1985** [reply]
> Perhaps you can use `MapAreaControllerRoute()` (see https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0 and example in the OC source) to add a higher prio route in your module/theme's `Startup`?
> 
> Also, @hishamco you can help, perhaps?

How can I remove the existing routes of `OrchardCore.Users`, like **/login**?

**Reply by hishamco** [reply]
This another thing, you can do if you replace the route in other module that's run after OC.Users module. FYI I remembered @hyzx86 did something similar in the past

**Reply by hyzx86** [reply]
Perhaps this is what you are referring to?
https://github.com/EasyOC/EasyOC.Modules/tree/main/src/Modules/EasyOC.ReplaceAction

It's just a replacement  the Action implementation on OrchardCore, nothing to do with routing.

**Reply by DotCat1985** [reply]
> It's just a replacement the Action implementation on OrchardCore, nothing to do with routing.

In the following sample, could I manipulate on action routes from `descriptor` object?

```
//sample 5 , Advanced configuration
    services.ReplaceAction(opt =>
    {
        opt.Items.Add(new ActionReplaceOptionItem
        {
            CustomAction = (descriptor) =>
            {
                //Do whatever you want
                Console.WriteLine(descriptor.ControllerName);
            }
        });
    });
```

**Reply by hyzx86** [reply]
You can use it to implement jumps, for example checking if the URL contains the specified culture, and if no culture option is found, jumping to the specified culture.

look at the OC source code for how to identify multi-tenancy from routing.

https://github.com/OrchardCMS/OrchardCore/blob/84ae5cb0c8e0ec297a645f32b39695c3ea7003a3/src/OrchardCore/OrchardCore/Modules/ModularTenantRouterMiddleware.cs#L28-L39

**Reply by hyzx86** [reply]
Pls Follow https://github.com/OrchardCMS/OrchardCore/issues/5125

**Comment by hishamco** [comment]
Why you didn't use localization routing which is feature that many might not now about, BTW similar issue already there

https://github.com/hishamco/aspnetcore/tree/main/src%2FMiddleware%2FLocalization.Routing

Hopefully I will reply with detailed answer tomorrow coz my Grandma is RIP today 😔

**Reply by DotCat1985** [reply]
I express my condolences and so you can reply me without any urgency.

**Reply by hishamco** [reply]
@DotCat1985 This might be useful https://github.com/OrchardCMS/OrchardCore/issues/5125
